restore runnability for pg < 9.4 - second try

eschen42 opened the pull request:
I was unable to run the image resulting from 
```bash
sudo docker-compose build pg9_3
```
This request has the fixes to support PostgreSQL versions < 9.4 which don't implement
- max_replication_slots in postgresql.conf
- pg_create_physical_replication_slot in the database

This is my second try at pull request, adjusted per comments at [pull request 1](https://github.com/tbeadle/postgres/pull/1).

Note that I did adjust `Dockerfile.template` to create the `/home/postgres` with the correct ownership and permissions to accept inbound SSH connections from the `postgres` user.  It is my understanding that it is somewhat customary to treat the `postgres` user as a system account and not create a `/home/postgres` directory, instead setting as home the default location for the database. However, in this example, the password file does rely upon `/home/postgres/.ssh/authorized_keys`, and I think it would be peculiar to put that file at `/var/lib/postgresql/data/.ssh/authorized_keys`. 

An additional side effect of not having `/home/postgres` owned by `postgres` is that, when `postgres` exits a `psql` session, there is an error message saying that `.psql_history` could not be written; it seems to me that setting the permissions properly makes more sense than forbidding the writing of the file unless that is a deliberate part of a security strategy. This example doesn't seem to me to be trying to model robust security but rather standard (and acceptable) practice.

If you prefer, I can remove the adjustment of ownership and perms of `/home/postgres`.

Thank you for this sample; it's a great starting point for experimentation.

eschen42 committed:
restore runnability for pg < 9.4 - second try

tbeadle commented on barman/image-pre-start.d/01_update_postgresql.conf.sh:
```diff
@@ -3,16 +3,22 @@
 set -eo pipefail
 
 . /usr/local/bin/functions.sh
+major_gte() { verlt $1 ${PG_MAJOR}; }
```
Please move this into functions.sh to keep it with all of the other version comparison functions.

eschen42 replied:
Are you confident that functions.sh would only be sourced when $PG_MAJOR is defined?

tbeadle replied:
Yes.  I believe so.  It's defined in the Dockerfile, so we should be good.

tbeadle commented on barman/Dockerfile.template:
```diff
@@ -6,7 +6,10 @@ RUN apt-get update \
 	&& rm -Rf /var/lib/apt/lists/* \
-	&& sed -i 's/\(.*pam_loginuid.so\)/#\1/' /etc/pam.d/sshd
+	&& sed -i 's/\(.*pam_loginuid.so\)/#\1/' /etc/pam.d/sshd \
+	&& mkdir /home/postgres \
```
Instead of manually doing these steps, just  add a `-m` argument to the `useradd` command in Dockerfile-debian.template.  That should cause /home/postgres to get created with the correct ownership/permissions.

eschen42 replied:
I can do that if you like, but I think that, ordinarily, people consuming 
`tbeadle/postgres/Dockerfile-debian.template`
may be expecting `postgres` to be a nonuser-id, as the current template suggests, which makes sense for the "use postgres without barman" use case.

It is only because `tbeadle/postgres/barman` relies upon`/home/postgres/.ssh/authorized_keys` that creating and using `/home/postgres` makes sense to me.  I still can make the change that you requested instead, but "are you sure?"  :)

tbeadle replied:
Good point!  Let's at least use `install -d -m 0700 -o postgres -g postgres /home/postgres` though, instead of the mkdir, chown, chmod combo.

tbeadle reviewed (changes requested):
Thanks again so much for your contributions here.

eschen42 committed:
Made changes pursuant to https://github.com/tbeadle/postgres/pull/2

specifically:
https://github.com/tbeadle/postgres/pull/2#discussion_r119090942
https://github.com/tbeadle/postgres/pull/2#discussion_r119091394

Validated that I could run psql as postgres and write the psql history
file to /home/postgres.

Built by running
  ./update.sh
  sudo docker-compose build pg9_3
in the main directory, then running
  ./update.sh
  sudo docker-compose build pg9_3
in the barman subdirectory, then running
  sudo docker run --rm -ti tbeadle/postgres:9.3-barman
and then
  su - postgres
  psql

	modified:   9.2/Dockerfile
	modified:   9.2/alpine/Dockerfile
	modified:   9.2/functions.sh
	modified:   9.3/Dockerfile
	modified:   9.3/alpine/Dockerfile
	modified:   9.3/functions.sh
	modified:   9.4/Dockerfile
	modified:   9.4/alpine/Dockerfile
	modified:   9.4/functions.sh
	modified:   9.5/Dockerfile
	modified:   9.5/alpine/Dockerfile
	modified:   9.5/functions.sh
	modified:   9.6/Dockerfile
	modified:   9.6/alpine/Dockerfile
	modified:   9.6/functions.sh
	modified:   Dockerfile-debian.template
	modified:   barman/Dockerfile-9.2
	modified:   barman/Dockerfile-9.3
	modified:   barman/Dockerfile-9.4
	modified:   barman/Dockerfile-9.5
	modified:   barman/Dockerfile-9.6
	modified:   barman/Dockerfile.template
	modified:   barman/image-pre-start.d/01_update_postgresql.conf.sh
	modified:   functions.sh

tbeadle commented on Dockerfile-debian.template:
```diff
@@ -2,7 +2,9 @@
-RUN groupadd -r postgres --gid=999 && useradd -r -g postgres --uid=999 postgres
+RUN groupadd -r postgres --gid=999 \
+  && useradd -r -g postgres --uid=999 postgres \
+  && install -d -m 0700 -o postgres -g postgres /home/postgres
```
I thought this was going to go in the barman/Dockerfile.template like you had originally done and not here so that the /home/postgres directory doesn't exist when not using barman.

eschen42 committed:
Updated files per pull-request discussion

Tested lightly:
  - ran initdb from bash
    - when logged in as postgres
    - for a directory under /home/postgres
  - validated that defalut db comes up when using docker run
    - in ..../postgres/barman ran
      - sudo docker-compose run pg9_3
    - in another window ran
      - sudo docker exec -u postgres -ti \
          barman_pg9_3_run_1 bash -c 'export TERM=xterm; /bin/bash'
      - from the bash prompt in the container (as postgres) ran
        - psql -c "select 1;"

On branch restore_pg9_3_try2

Changes as requested in discussion of pull request
 	modified:   ../9.2/Dockerfile
 	modified:   ../9.3/Dockerfile
 	modified:   ../9.4/Dockerfile
 	modified:   ../9.5/Dockerfile
 	modified:   ../9.6/Dockerfile
 	modified:   ../Dockerfile-debian.template
Reversion to original (up to whitespace changes at least)
 	modified:   Dockerfile-9.2
 	modified:   Dockerfile-9.3
 	modified:   Dockerfile-9.4
 	modified:   Dockerfile-9.5
 	modified:   Dockerfile-9.6
 	modified:   Dockerfile.template

eschen42 commented:
This is working for me with pg 9.3.  I'm not sure why the third version number is changed in the non-barman Dockerfiles - could it be that this is an artifact of running postgres/update.sh ?
